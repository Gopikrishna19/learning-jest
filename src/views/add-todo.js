import PropTypes from 'prop-types';
import React, {Component} from 'react';

export class AddTodo extends Component {
    constructor() {
        super();
        this.state = {todo: ''};
    }

    submitOnEnter = event => event.key === 'Enter' && this.handleClick();
    handleChange = event => this.setState({todo: event.target.value});
    handleClick = () => {
        if (this.state.todo.replace(/\s/g, '')) {
            this.props.onAdd(this.state.todo);
            this.setState({todo: ''});
        }
    };

    render() {
        return (
            <div>
                <input
                    autoFocus={true}
                    onKeyPress={this.submitOnEnter}
                    onChange={this.handleChange}
                    type='text'
                    value={this.state.todo}
                />
                <button onClick={this.handleClick}>Add</button>
            </div>
        );
    }

}

AddTodo.displayName = 'AddTodo';
AddTodo.propTypes = {
    onAdd: PropTypes.func.isRequired
};
