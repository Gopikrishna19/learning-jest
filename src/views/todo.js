import PropTypes from 'prop-types';
import React from 'react';
import {todoFinished} from './todo.scss';

const handleChange = props => () => props.onToggle(!props.todo.finished);

const getClassNames = props => [
    'todo',
    props.todo.finished ? todoFinished : ''
]
    .filter(name => name)
    .join(' ');

export const Todo = props =>
    <div className={getClassNames(props)}>
        <input
            onChange={handleChange(props)}
            checked={props.todo.finished}
            type='checkbox'
        />
        {props.todo.todo}
        <button onClick={props.onRemove}>Remove</button>
    </div>;

Todo.displayName = 'Todo';
Todo.propTypes = {
    onRemove: PropTypes.func.isRequired,
    onToggle: PropTypes.func.isRequired,
    todo: PropTypes.object.isRequired
};
