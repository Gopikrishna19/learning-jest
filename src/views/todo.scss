.todoFinished {
    text-decoration: line-through;
    color: #333;
}
