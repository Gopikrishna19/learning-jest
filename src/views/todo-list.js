import PropTypes from 'prop-types';
import React from 'react';
import {Todo} from './todo';

const handleRemove = (props, todo) => () => props.removeTodo(todo.id);
const handleToggle = (props, todo) => selection => props.toggleTodo(todo.id, selection);

export const TodoList = props =>
    <div className='todo-list'>
        {
            props.todos.map(
                todo =>
                    <Todo
                        todo={todo}
                        key={todo.id}
                        onRemove={handleRemove(props, todo)}
                        onToggle={handleToggle(props, todo)}
                    />
            )
        }
    </div>;

TodoList.displayName = 'TodoList';
TodoList.propTypes = {
    removeTodo: PropTypes.func.isRequired,
    todos: PropTypes.array.isRequired,
    toggleTodo: PropTypes.func.isRequired
};
