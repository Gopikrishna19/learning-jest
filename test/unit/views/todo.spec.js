import {shallow} from 'enzyme';
import React from 'react';
import {Todo} from '../../../src/views/todo';

describe('<Todo/>', () => {
    let todoElement,
        testProps;

    beforeEach(() => {
        testProps = {
            onRemove: jest.fn(() => null),
            onToggle: jest.fn(() => null),
            todo: {
                finished: false,
                todo: 'to buy milk'
            }
        };
        todoElement = shallow(<Todo {...testProps}/>);
    });

    it('should be a div with a class name', () => {
        expect(todoElement.getElement()).toEqual(
            <div className='todo'>
                {expect.anything()}
            </div>
        );
    });

    it('should have checkbox to toggle todo', () => {
        const input = todoElement.childAt(0);

        expect(input.getElement()).toEqual(
            <input
                onChange={expect.any(Function)}
                checked={testProps.todo.finished}
                type='checkbox'
            />
        );

        input.props().onChange();

        expect(testProps.onToggle).toHaveBeenCalledTimes(1);
        expect(testProps.onToggle).toHaveBeenCalledWith(!testProps.todo.finished);
    });

    it('should have todo text', () => {
        expect(todoElement.childAt(1).text()).toBe(testProps.todo.todo);
    });

    it('should have a button to remove the todo', () => {
        expect(todoElement.childAt(2).getElement()).toEqual(
            <button onClick={testProps.onRemove}>Remove</button>
        );
    });

    describe('after finishing the todo', () => {
        beforeEach(() => {
            testProps.todo.finished = true;
            todoElement.setProps(testProps);
        });

        it('should be a div with a class names', () => {
            expect(todoElement.getElement()).toEqual(
                <div className='todo todoFinished'>
                    {expect.anything()}
                </div>
            );
        });
    });
});
