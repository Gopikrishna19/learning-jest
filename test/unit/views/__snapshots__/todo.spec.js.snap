// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo/> after finishing the todo should match snapshot after finishing: after-finishing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todo
    onRemove={[MockFunction]}
    onToggle={[MockFunction]}
    todo={
      Object {
        "finished": true,
        "todo": "to buy milk",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={true}
          onChange={[Function]}
          type="checkbox"
        />,
        "to buy milk",
        <button
          onClick={[MockFunction]}
        >
          Remove
        </button>,
      ],
      "className": "todo todoFinished",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": true,
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "to buy milk",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Remove",
          "onClick": [MockFunction],
        },
        "ref": null,
        "rendered": "Remove",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={true}
            onChange={[Function]}
            type="checkbox"
          />,
          "to buy milk",
          <button
            onClick={[MockFunction]}
          >
            Remove
          </button>,
        ],
        "className": "todo todoFinished",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": true,
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "to buy milk",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Remove",
            "onClick": [MockFunction],
          },
          "ref": null,
          "rendered": "Remove",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<Todo/> should match snapshot: before-finishing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todo
    onRemove={[MockFunction]}
    onToggle={[MockFunction]}
    todo={
      Object {
        "finished": false,
        "todo": "to buy milk",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />,
        "to buy milk",
        <button
          onClick={[MockFunction]}
        >
          Remove
        </button>,
      ],
      "className": "todo",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": false,
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "to buy milk",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Remove",
          "onClick": [MockFunction],
        },
        "ref": null,
        "rendered": "Remove",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />,
          "to buy milk",
          <button
            onClick={[MockFunction]}
          >
            Remove
          </button>,
        ],
        "className": "todo",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": false,
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "to buy milk",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Remove",
            "onClick": [MockFunction],
          },
          "ref": null,
          "rendered": "Remove",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
